File: scripts/fix_dual_gender.py
#!/usr/bin/env python3
import json, sys
sys.stdout.reconfigure(encoding='utf-8')

d = json.load(open('data/dual-gender-nouns.json', encoding='utf-8'))

fixes = {
    'Mensch': {'der': 'human/person', 'das': 'wretch/hussy (derogatory)'},
    'Tag': {'der': 'day', 'das': 'tag (markup)'},
    'Teil': {'der': 'part/portion', 'das': 'piece/component'},
    'Kunde': {'die': 'news/tidings', 'der': 'customer/client'},
    'Ort': {'der': 'place/location', 'das': 'pointed end/awl'},
    'Fall': {'der': 'case/fall', 'das': 'halyard (nautical rope)'},
    'Alter': {'der': 'old man (colloquial)', 'das': 'age'},
    'Service': {'der': 'service', 'das': 'tea/dinner set'},
    'Bund': {'der': 'federation/alliance', 'das': 'bunch/bundle'},
    'Band': {'die': 'band (music group)', 'der': 'volume (book)', 'das': 'ribbon/tape/bond'},
    'See': {'die': 'sea/ocean', 'der': 'lake'},
    'Moment': {'der': 'moment (time)', 'das': 'factor/element (physics: torque)'},
    'Leiter': {'die': 'ladder', 'der': 'leader/director/conductor'},
    'Maß': {'die': 'liter of beer (Bavarian)', 'das': 'measure/degree/extent'},
    'Tor': {'der': 'fool (archaic)', 'das': 'gate/goal'},
    'Kaffee': {'der': 'coffee', 'das': 'cafe'},
    'Steuer': {'die': 'tax', 'das': 'steering wheel/helm'},
    'Post': {'die': 'mail/post office', 'der': 'post (online)'},
    'Morgen': {'der': 'morning', 'das': 'tomorrow/the future'},
    'Mangel': {'die': 'mangle (laundry press)', 'der': 'lack/deficiency'},
    'Gang': {'die': 'gang', 'der': 'corridor/aisle/gear/gait'},
    'Gehalt': {'der': 'content/substance', 'das': 'salary/wage'},
    'Bord': {'der': 'board (on board)', 'das': 'shelf'},
    'Boot': {'der': 'boot (footwear, from English)', 'das': 'boat'},
    'Erbe': {'der': 'heir', 'das': 'inheritance/heritage'},
    'Mode': {'die': 'fashion', 'der': 'mode (technical)'},
    'Bauer': {'der': 'farmer/peasant', 'das': 'birdcage'},
    'Loch': {'der': 'loch (Scottish lake)', 'das': 'hole'},
    'Tee': {'der': 'tea', 'das': 'tee (golf)'},
    'Chor': {'der': 'choir/chorus', 'das': 'chancel (church architecture)'},
    'Mund': {'die': 'guardianship (legal, archaic)', 'der': 'mouth'},
    'Kraut': {'der': 'Kraut (derogatory)', 'das': 'herb/cabbage'},
    'Schild': {'der': 'shield', 'das': 'sign/nameplate'},
    'Wende': {'die': 'turn/change (die Wende = reunification)', 'der': 'Wend/Sorb (Slavic people)'},
    'Model': {'der': 'mold/form', 'das': 'model (fashion)'},
    'Plastik': {'die': 'sculpture', 'der': 'plastic'},
    'Mark': {'die': 'mark (currency)/border march', 'das': 'marrow/pith'},
    'Verdienst': {'der': 'earnings/income', 'das': 'merit/credit'},
    'Lob': {'der': 'lob (sports)', 'das': 'praise'},
    'Flur': {'die': 'open field/farmland', 'der': 'hallway/corridor'},
    'Stift': {'der': 'pen/pencil/apprentice', 'das': 'collegiate church/foundation'},
    'Messer': {'der': 'measurer/gauge', 'das': 'knife'},
    'Golf': {'der': 'gulf', 'das': 'golf (sport)'},
    'Militaer': {'der': 'military officer', 'das': 'military/army'},
    'Militär': {'der': 'military officer', 'das': 'military/army'},
    'Reis': {'der': 'rice', 'das': 'twig/shoot'},
    'Schock': {'der': 'shock', 'das': 'sixty (archaic unit)'},
    'Objektiv': {'der': 'objective case (grammar)', 'das': 'lens/objective (camera)'},
    'Gift': {'die': 'gift (archaic/false friend)', 'der': 'anger/wrath (archaic)', 'das': 'poison'},
    'Gefallen': {'der': 'favour', 'das': 'pleasure/liking'},
    'Heide': {'die': 'heath/moor', 'der': 'heathen/pagan'},
    'Kiefer': {'die': 'pine tree', 'der': 'jaw'},
    'Mast': {'die': 'fattening (of animals)', 'der': 'mast/pole/pylon'},
    'Korn': {'der': 'grain schnapps', 'das': 'grain/corn/kernel'},
    'Pony': {'der': 'fringe/bangs (hair)', 'das': 'pony (small horse)'},
    'Marsch': {'die': 'marsh/wetland', 'der': 'march'},
    'Bulle': {'die': 'papal bull', 'der': 'bull/cop (slang)'},
    'Lama': {'der': 'lama (Buddhist monk)', 'das': 'llama (animal)'},
    'Laster': {'der': 'lorry/truck', 'das': 'vice/bad habit'},
    'Rentier': {'der': 'person of independent means', 'das': 'reindeer'},
    'Titan': {'der': 'Titan (mythology)', 'das': 'titanium (element)'},
    'Heroin': {'die': 'heroine', 'das': 'heroin (drug)'},
    'Page': {'die': 'page (web/book)', 'der': 'bellboy/page (attendant)'},
    'Otter': {'die': 'adder/viper', 'der': 'otter'},
    'Orange': {'die': 'orange (fruit)', 'das': 'orange (color)'},
    'Ekel': {'der': 'disgust/revulsion', 'das': 'disgusting person'},
    'Ohm': {'der': 'uncle (dialectal)', 'das': 'ohm (unit)'},
    'Tau': {'der': 'dew', 'das': 'rope/hawser'},
    'Pik': {'der': 'grudge/pique', 'das': 'spades (card suit)'},
    'Zimt': {'der': 'nonsense (colloquial)', 'das': 'cinnamon'},
    'Zink': {'der': 'cornett (Renaissance instrument)', 'das': 'zinc (metal)'},
    'Python': {'der': 'python (snake)', 'das': 'Python (programming)'},
    'Espresso': {'der': 'espresso (coffee)', 'das': 'espresso bar'},
    'Franchise': {'die': 'deductible/excess (insurance)', 'das': 'franchise'},
    'Spektakel': {'der': 'racket/commotion', 'das': 'spectacle/show'},
    'Hut': {'der': 'hat', 'die': 'guard/protection'},
}

for noun in d['nouns']:
    word = noun['word']
    if word in fixes:
        for g_entry in noun['genders']:
            art = g_entry['article']
            if art in fixes[word]:
                g_entry['english'] = fixes[word][art]

with open('data/dual-gender-nouns.json', 'w', encoding='utf-8') as f:
    json.dump(d, f, ensure_ascii=False, indent=2)

print(f'Fixed translations for {len(d["nouns"])} dual-gender nouns.')
print()
for noun in d['nouns']:
    parts = [f'{g["article"]} = {g["english"]}' for g in noun['genders']]
    print(f'  #{noun.get("rank","?"):<5} {noun["word"]:20s} {"  |  ".join(parts)}')

File: scripts/fix-dual-gender.js
#!/usr/bin/env node
/**
 * Fix known errors in dual-gender-nouns.json identified during audit.
 * Run: node scripts/fix-dual-gender.js
 */

const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, '..', 'data', 'dual-gender-nouns.json');
const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));

let fixCount = 0;

function findByWord(word) {
  return data.nouns.find(n => n.word === word);
}

// ============================================================
// 1. FIX WRONG ARTICLES
// ============================================================
console.log('\n=== FIXING WRONG ARTICLES ===');

// Plastik: der Plastik (plastic material) should be das Plastik
const plastik = findByWord('Plastik');
if (plastik) {
  for (const g of plastik.genders) {
    if (g.article === 'der' && g.english.includes('plastic')) {
      console.log(`[FIX] "Plastik": "der" (plastic) -> "das"`);
      g.article = 'das';
      fixCount++;
    }
  }
}

// Alter: der Alte plural should be "Alten" not "Alte"
const alter = findByWord('Alter');
if (alter) {
  for (const g of alter.genders) {
    if (g.article === 'der' && g.plural === 'Alte') {
      console.log(`[FIX] "Alter": der Alte plural "Alte" -> "Alten"`);
      g.plural = 'Alten';
      fixCount++;
    }
  }
}

// Mensch: das Mensch plural "Menscher" — flag as archaic/dialectal
const mensch = findByWord('Mensch');
if (mensch) {
  for (const g of mensch.genders) {
    if (g.article === 'das') {
      g.english = 'wretch (archaic/dialectal, derogatory)';
      console.log(`[FIX] "Mensch": das Mensch — clarified as archaic/dialectal`);
      fixCount++;
    }
  }
}

// ============================================================
// 2. REMOVE QUESTIONABLE ENTRIES
// ============================================================
console.log('\n=== REMOVING QUESTIONABLE ENTRIES ===');

const removeWords = [
  'Loch',      // "der Loch" = Scottish lake — not German
  'Kraut',     // "der Kraut" = derogatory term — not a real noun gender distinction
];

const beforeCount = data.nouns.length;
data.nouns = data.nouns.filter(n => {
  if (removeWords.includes(n.word)) {
    console.log(`[REMOVE] "${n.word}": not a genuine German dual-gender noun`);
    fixCount++;
    return false;
  }
  return true;
});

// ============================================================
// 3. FIX MISLEADING ENTRIES
// ============================================================
console.log('\n=== FIXING MISLEADING ENTRIES ===');

// Kaffee: das Kaffee = cafe is dialectal, update to note this
const kaffee = findByWord('Kaffee');
if (kaffee) {
  for (const g of kaffee.genders) {
    if (g.article === 'das') {
      g.english = 'cafe (regional/dialectal, standard: das Café)';
      console.log(`[FIX] "Kaffee": das Kaffee marked as regional/dialectal`);
      fixCount++;
    }
  }
}

// Espresso: das Espresso = espresso bar is non-standard
const espresso = findByWord('Espresso');
if (espresso) {
  for (const g of espresso.genders) {
    if (g.article === 'das') {
      g.english = 'espresso (variant gender, non-standard)';
      console.log(`[FIX] "Espresso": das Espresso marked as non-standard`);
      fixCount++;
    }
  }
}

// Reis: plural "Reise" is confusing (means "journey")
const reis = findByWord('Reis');
if (reis) {
  for (const g of reis.genders) {
    if (g.article === 'der' && g.plural === 'Reise') {
      g.plural = '—';
      g.singularOnly = true;
      console.log(`[FIX] "Reis": der Reis (rice) — marked as mass noun, no standard plural`);
      fixCount++;
    }
  }
}

// ============================================================
// 4. ADD MISSING IMPORTANT ENTRIES
// ============================================================
console.log('\n=== ADDING MISSING ENTRIES ===');

// Check if Hut already exists
if (!findByWord('Hut')) {
  data.nouns.push({
    word: 'Hut',
    genders: [
      {
        article: 'der',
        english: 'hat',
        plural: 'Hüte',
        singularOnly: false,
        pluralOnly: false
      },
      {
        article: 'die',
        english: 'guard/protection (auf der Hut sein = to be on guard)',
        plural: '—',
        singularOnly: true,
        pluralOnly: false
      }
    ],
    rank: 999
  });
  console.log(`[ADD] "Hut": der Hut (hat) / die Hut (guard)`);
  fixCount++;
}

// Check if Junge already exists
if (!findByWord('Junge')) {
  data.nouns.push({
    word: 'Junge',
    genders: [
      {
        article: 'der',
        english: 'boy',
        plural: 'Jungen',
        singularOnly: false,
        pluralOnly: false
      },
      {
        article: 'das',
        english: 'young animal/cub',
        plural: 'Jungen',
        singularOnly: false,
        pluralOnly: false
      }
    ],
    rank: 998
  });
  console.log(`[ADD] "Junge": der Junge (boy) / das Junge (young animal)`);
  fixCount++;
}

// Update meta count
data.meta.count = data.nouns.length;

// ============================================================
// WRITE OUTPUT
// ============================================================
console.log(`\n=== TOTAL FIXES: ${fixCount} ===`);
fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf8');
console.log(`Written to ${filePath}`);
